File: src/geometry/calculate.f90
module calculate
use math
use geom_utils
use divide
implicit none

private
public :: triangle_area
public :: cp_lengths, cp_net_areas
public :: arc_length_bezier, surface_area_bezier

contains

function triangle_area(m, tri) result(area)
    !> Calculate the area of a triangle.
    !> m - Number columns in triangle array.
    !> tri - Triangle array.
    !> area - Area of triangle.
    
    ! Input
    integer, intent(in) :: m
    double precision, intent(in) :: tri(3, m)
    
    ! Output
    double precision :: area
    
    ! Working
    double precision :: mag
    double precision :: v1(m), v2(m), vc(m)
    
    v1 = tri(2, :) - tri(1, :)
    v2 = tri(3, :) - tri(1, :)
    
    vc = cross(v1, v2)
    mag = norm(vc)
    area = 0.50d0 * mag
    
end function triangle_area

subroutine cp_lengths(n, cp, lp, lc)
    !> Calculate the linear length of the control polygon and its chord.
    !> n - Number of control points - 1.
    !> cp - Dehomogenized control points.
    !> lp - Length of control polygon.
    !> lc - Chord length.
    
    !f2py intent(in) n, cp
    !f2py intent(out) lp, lc
    !f2py depend(n) cp
    
    ! Input
    integer, intent(in) :: n
    
    ! Output
    double precision, intent(out) :: lp, lc
    double precision, intent(in) :: cp(0:n, 0:2)    
    
    ! Working
    integer :: i
    double precision :: vec(0:2)
    
    vec = cp(n, :) - cp(0, :)
    lc = norm(vec)
    lp = 0.0d0
    do i = 0, n - 1
        vec = cp(i + 1, :) - cp(i, :)
        lp = lp + norm(vec)
    end do
    
end subroutine cp_lengths

subroutine cp_net_areas(n, m, cp, anet, ap)
    !> Calculate the surface areas of the control net.
    !> n - Number of control points - 1 in u-direction.
    !> m - Number of control points - 1 in v-direction.
    !> cp - Dehomogenized control points.
    !> anet - Area of control net.
    !> ap - Area of the plane formed by the four corners of the
    !> control net.
    
    !f2py intent(in) n, m, cp
    !f2py intent(out) anet, ap
    !f2py depend(n, m) cp
    
    ! Input
    integer, intent(in) :: n, m
    double precision, intent(in) :: cp(0:n, 0:m, 0:2)
    
    ! Output
    double precision, intent(out) :: anet, ap
    
    ! Working
    integer :: i, ntri
    integer :: triangles(0:2 * n * m - 1, 0:2)
    double precision :: v1(0:2), v2(0:2), v3(0:2)
    double precision :: verts(0:(n + 1) * (m + 1) - 1, 0:2)
    double precision :: c1(0:2), c2(0:2), c3(0:2)
                        
    call tessellate_cp_net(n, m, cp, verts, triangles)
    ntri = size(triangles, 1)
    anet = 0.0d0
    do i = 0, ntri - 1
        v1 = verts(triangles(i, 0), :)
        v2 = verts(triangles(i, 1), :)
        v3 = verts(triangles(i, 2), :)
        c1 = v2 - v1
        c2 = v3 - v1
        c3 = cross(c1, c2)
        anet = anet + 0.50d0 * norm(c3)
    end do
    c1 = cp(n, 0, :) - cp(0, 0, :)
    c2 = cp(n, m, :) - cp(0, 0, :)
    c3 = cross(c1, c2)
    ap = 0.50d0 * norm(c3)
    c1 = cp(n, m, :) - cp(0, 0, :)
    c2 = cp(0, m, :) - cp(0, 0, :)
    c3 = cross(c1, c2)
    ap = ap + 0.50d0 * norm(c3)
end subroutine cp_net_areas

recursive subroutine arc_length_bezier(n, cpw, u0, u1, tol, l)
    !> Estimate the arc length of a Bezier curve.
    !> n - Number of control points - 1.
    !> cpw - Control points.
    !> u0 - Starting parameter.
    !> u1 - Ending parameter.
    !> tol - Tolerance for comparing control polygon length and chord
    !> length.
    !> l - Approximate arc length.
    
    !f2py intent(in) n, cpw, tol, u0, u1
    !f2py intent(in,out) l
    !f2py depend(n) cpw
    
    ! Input
    integer, intent(in) :: n
    double precision, intent(in) :: u0, u1, tol
    double precision, intent(in) :: cpw(0:n, 0:3)
    
    ! Input/Output
    double precision, intent(inout) :: l(1)
    
    ! Working
    integer :: i
    double precision :: lp, lc, a1, b1, a2, b2
    double precision :: cp(0:n, 0:2)
    double precision :: qw1(0:n, 0:3), qw2(0:n, 0:3)
    double precision :: temp2(0:n, 0:3), w(0:n)
                        
    ! Extract curve if needed
    if ((u0 .gt. 0.0d0) .or. (u1 .lt. 1.0d0)) then
        call extract_bezier_curve(cpw, n, u0, u1, 0.0d0, 1.0d0, temp2, a1, b1)
    else
        temp2 = cpw
    end if

    call dehomogenize_array1d(n, temp2, cp, w)
    call cp_lengths(n, cp, lp, lc)
    
    if (dabs(lp - lc) .le. tol) then
        l(1) = l(1) + (2.0d0 * lc + (dble(n) - 1.0d0) * lp) / (dble(n) + 1.0d0)
    else
        call split_bezier_curve(temp2, n, 0.50d0, 0.0d0, 1.0d0, qw1, a1, b1, &
                                qw2, a2, b2)
        call arc_length_bezier(n, qw1, 0.0d0, 1.0d0, tol, l)
        call arc_length_bezier(n, qw2, 0.0d0, 1.0d0, tol, l)
    end if
    
end subroutine arc_length_bezier

recursive subroutine surface_area_bezier(n, m, cpw, tol, a)
    !> Estimate the surface area of a Bezier surface.
    !> n - Number of control points - 1 in u-direction.
    !> m - Number of control points - 1 in v-direction.
    !> cpw - Control points.
    !> tol - Tolerance for comparing area of control net to area of
    !> base area of formed by the four corners of the control net.
    !> a - Approximate area of surface.
    
    !f2py intent(in) n, m, cpw, tol
    !f2py intent(in,out) a
    !f2py depend(n, m) cpw
    
    ! Input 
    integer, intent(in) :: n, m
    double precision, intent(in) :: tol
    double precision, intent(in) :: cpw(0:n, 0:m, 0:3)
    
    ! Input/Output
    double precision, intent(inout) :: a(1)
    
    integer :: i
    double precision :: ap, anet
    double precision :: cp(0:n, 0:m, 0:2)
    double precision :: qw1(0:n, 0:m, 0:3), qw2(0:n, 0:m, 0:3)
    double precision :: qw3(0:n, 0:m, 0:3), qw4(0:n, 0:m, 0:3)
    double precision :: qw5(0:n, 0:m, 0:3), qw6(0:n, 0:m, 0:3)
    double precision :: w(0:n, 0:m)
    double precision :: ab1(0:1, 0:1), ab2(0:1, 0:1)

    call dehomogenize_array2d(n, m, cpw, cp, w)
    call cp_net_areas(n, m, cp, anet, ap)
    
    if (abs(anet - ap) .le. tol) then
        a(1) = a(1) + (4.0d0 * ap + dble(n + m + n * m - 3) * anet) / &
               dble((n + 1) * (m + 1))
    else
        call split_bezier_surface(cpw, n, m, 0.50d0, -1.0d0, &
                                  0.0d0, 1.0d0, 0.0d0, 1.0d0, &
                                  qw1, qw2, ab1, ab2)
        call split_bezier_surface(qw1, n, m, -1.0d0, 0.50d0, &
                                  0.0d0, 1.0d0, 0.0d0, 1.0d0, &
                                  qw3, qw4, ab1, ab2)
        call split_bezier_surface(qw2, n, m, -1.0d0, 0.50d0, &
                                  0.0d0, 1.0d0, 0.0d0, 1.0d0, &
                                  qw5, qw6, ab1, ab2)
        call surface_area_bezier(n, m, qw3, tol, a)
        call surface_area_bezier(n, m, qw4, tol, a)
        call surface_area_bezier(n, m, qw5, tol, a)
        call surface_area_bezier(n, m, qw6, tol, a)
    end if
    
end subroutine surface_area_bezier

end module calculate

File: pynurbs/geometry/methods/calculate.py
from __future__ import division

from numpy import array, cross, diff, float64
from numpy.linalg import norm

from pynurbs.config import Settings
from pynurbs.geometry.methods.divide import (extract_bezier_curve,
                                             extract_nurbs_curve,
                                             split_bezier_curve,
                                             split_bezier_surface)
from pynurbs.geometry.methods.geom_utils import (dehomogenize_array1d,
                                                 dehomogenize_array2d,
                                                 tessellate_cp_net)
from pynurbs.geometry.methods.modify import decompose_curve


def polygon_area2D(xy):
    """
    Calculate the signed area of a 2D (planar) polygon using the shoelace
    formula.

    :param array_like xy: Ordered array of points defining polygon boundary.

    :return: Area of polygon.
    :rtype: float
    """
    xy = array(xy, dtype=float64)
    nm = xy.shape

    x0, y0 = xy[0]
    xn, yn = xy[-1]
    a = xn * y0 - x0 * yn
    for i in range(0, nm[0] - 1):
        a += xy[i, 0] * xy[i + 1, 1]
        a -= xy[i + 1, 0] * xy[i, 1]
    a *= 0.5

    return a


def triangle_area(t):
    """
    Calculate the area of a triangle given by 3 points.

    :param array_like t: Array containing triangle vertices.
    """
    t = array(t, dtype=float64)
    return 0.5 * norm(cross(t[1] - t[0], t[2] - t[0]))


def cp_lengths(n, cp):
    """
    Calculate the linear length of the control polygon and its chord.

    :param int n: Number of control points - 1.
    :param ndarray cp: Dehomogenized control points.

    :return: Linear length of control polygon and its chord (lp, lc)
    :rtype: tuple
    """
    lp = sum(norm(diff(cp, axis=0), axis=1))
    lc = norm(cp[-1] - cp[0])
    return lp, lc


def cp_net_areas(n, m, cp):
    """
    Calculate the surface areas of the control point net by tessellating the
    control point net into triangles.

    :param int n: Control points - 1 in u-direction.
    :param int m: Control points - 1 in v-direction.
    :param ndarray cp: Dehomogenized control points.

    :return: Surface area of tessellated control point net and the area of the
        base plane (anet, ap).
    :rtype: tuple

    **Note:** The *base area* if the area of the two triangles formed by the
    four corner points of the control net.
    """
    # Tessellate the surface.
    verts, triangles = tessellate_cp_net(n, m, cp)
    ntri = triangles.shape[0]
    # Area of control net.
    anet = 0.
    for i in range(0, ntri):
        v1 = verts[triangles[i, 0]]
        v2 = verts[triangles[i, 1]]
        v3 = verts[triangles[i, 2]]
        anet += 0.5 * norm(cross(v2 - v1, v3 - v1))
    # Area of base plane.
    ap = (0.5 * norm(cross(cp[n, 0] - cp[0, 0], cp[n, m] - cp[0, 0])) +
          0.5 * norm(cross(cp[n, m] - cp[0, 0], cp[0, m] - cp[0, 0])))
    return anet, ap


def arc_length_bezier(cpw, n, u0=0., u1=1.):
    """
    Estimate the arc length of a Bezier curve.

    :param ndarray cpw: Control points.
    :param int n: Number of control points - 1 (degree).
    :param float u0: Starting parameter.
    :param float u1: Ending parameter.

    :return: Approximate arc length.
    :rtype: float
    """
    gtol = Settings.gtol
    arc_length = array([0.], dtype=float64)

    # Extract curve between u0 and u1 if needed.
    if u0 > 0. or u1 < 1.:
        cpw, _, _ = extract_bezier_curve(cpw, n, u0, u1)

    def _arc_length(cpw_i):
        cp, _ = dehomogenize_array1d(n, cpw_i)
        lp, lc = cp_lengths(n, cp)
        if abs(lp - lc) <= gtol:
            arc_length[0] += (2. * lc + (n - 1.) * lp) / (n + 1.)
        else:
            aw, _, _, bw, _, _ = split_bezier_curve(cpw_i, n, 0.5)
            _arc_length(aw)
            _arc_length(bw)

    _arc_length(cpw)
    return arc_length[0]


def surface_area_bezier(n, m, cpw):
    """
    Estimate the surface area of Bezier surface.

    :param int n: Number of control points - 1 in u-direction.
    :param int m: Number of control points - 1 in v-direction.
    :param ndarray cpw: Control points.

    :return: Approximate area of surface.
    :rtype: float
    """
    gtol = Settings.gtol
    surface_area = array([0.], dtype=float64)

    def _area(cpw_i):
        cp, _ = dehomogenize_array2d(n, m, cpw_i)
        anet, ap = cp_net_areas(n, m, cp)
        if abs(anet - ap) <= gtol:
            surface_area[0] += (4. * ap + (n + m + n * m - 3) * anet) / (
                    (n + 1) * (m + 1))
        else:
            qw1, qw2, _, _ = split_bezier_surface(cpw_i, n, m, u=0.5)
            qw3, qw4, _, _ = split_bezier_surface(qw1, n, m, v=0.5)
            qw5, qw6, _, _ = split_bezier_surface(qw2, n, m, v=0.5)
            _area(qw3)
            _area(qw4)
            _area(qw5)
            _area(qw6)

    _area(cpw)
    return surface_area[0]


def arc_length_nurbs(n, p, uk, cpw, u0=0., u1=1., a=0., b=1.):
    """
    Estimate the arc length of a NURBS curve between parameters *u0* and *u1*.

    :param int n: Number of control points - 1.
    :param int p: Degree.
    :param ndarray uk: Knot vector.
    :param ndarray cpw: Control points.
    :param float u0: Starting parameter.
    :param float u1: Ending parameter.
    :param float a: Lower domain of curve.
    :param float b: Upper domain of curve.

    :return: Estimated arc length.
    :rtype: float
    """
    # Extract curve between u0 and u1 if needed.
    if u0 > a or u1 < b:
        uk, cpw = extract_nurbs_curve(n, p, uk, cpw, u0, u1)
        n = cpw.shape[0] - 1
    # Decompose into Bezier segments
    nb, qw, _ = decompose_curve(n, p, uk, cpw)
    # Find arc length of each segment.
    d = 0.
    for i in range(nb):
        d += arc_length_bezier(qw[i], p)
    return d
